com.ibm.streamsx.topology/opt/python/packages/streamsx/rest_primitives.py

```diff
@@ -137,17 +139,105 @@ def __init__(self, username, password):
         self.session = requests.Session()
         self.session.auth = (username, password)
 
+    def handle_http_errors(self, res):
+        # HTTP error responses are 4xx, server errors are 5xx
+        if res.status_code >= 400:
+            logger.error("Response returned with error code: " + str(res.status_code))
+            logger.error(res.text)
+            res.raise_for_status()
+
     def make_request(self, url):
         logger.debug('Beginning a REST request to: ' + url)
-        return self.session.get(url).json()
+        res = self.session.get(url)
+        self.handle_http_errors(res)
+        return res.json()
+
+    def make_raw_request(self, url):
+        logger.debug('Beginning a REST request to: ' + url)
+        res = self.session.get(url)
+        self.handle_http_errors(res)
+        return res
+
+    def make_raw_streaming_request(self, url):
+        logger.debug('Beginning a REST request to: ' + url)
+        res = self.session.get(url, stream=True)
+        self.handle_http_errors(res)
+
+        return res
+
+    def __str__(self):
+        return pformat(self.__dict__)
+
+
+
+class _IAMStreamsRestClient(_StreamsRestClient):
+    """Handles the session connection with the Streams REST API and Streaming Analytics service
+    using IAM authentication.
+    """
+    def __init__(self, credentials):
+        """
+        Args:
+            credentials: The credentials of the Streaming Analytics service.
+        """
+        self._credentials = credentials
+        self._api_key = self._credentials[_IAMConstants.API_KEY]
+
+        # Represents the epoch time at which the token is no longer valid
+        # Starts at -1 such that the first invocation of a REST request
+        # Retrieves a token
+        self._auth_expiry_time = -1
+
+        # Determine if service is in stage1
+        if 'stage1' in  self._credentials[_IAMConstants.V2_REST_URL]:
+            self._token_url = _IAMConstants.TOKEN_URL_STAGE1
+        else:
+            self._token_url = _IAMConstants.TOKEN_URL
+
+        self.session = requests.Session()
+
+    def _get_authorization(self):
+        cur_time = int(time.time())
+        if cur_time >= self._auth_expiry_time:
+            self._refresh_authorization()
+        return self._bearer_token
+
+    def _refresh_authorization(self):
+        post_url = self._token_url + '?' + self._get_token_params(self._api_key)
+        res = requests.post(post_url, headers = {'Accept' : 'application/json',
+                                                 'Content-Type' : 'application/x-www-form-urlencoded'})
+        self.handle_http_errors(res)
+        res = res.json()
+
+        self._auth_exporiry_time = int(res[_IAMConstants.EXPIRATION]) - _IAMConstants.EXPIRY_PAD_MS
```

Comment:
Typo! Needs to be `_auth_expiry_time`

Reply:
Is there a mismatch of units here?
Above (line 199) `cur_time` represents the number of seconds since the epoch, but in setting the new expiry time "MS" is being used, which seems to be milliseconds.

Reply:
Yes, cur_time needs to be converted to ms.

Reply:
Fixed.